File: built-in plugins dsp/mb-filter-autowah.cpp
/**
 * MB Auto-Wah
 * Category : effect
 * Type     : eq
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Envelope-following auto-wah effect
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_FILTER_AUTOWAH_H
#define MB_FILTER_AUTOWAH_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbFilterAutowah : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-filter-autowah";
    static constexpr const char* PLUGIN_NAME    = "MB Auto-Wah";
    static constexpr const char* PLUGIN_TYPE    = "eq";
    static constexpr const char* PLUGIN_CATEGORY = "effect";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float sensitivity = 0.5f;  // range [0, 1]
    float speed = 0.5f;  // range [0, 1]
    float depth = 0.7f;  // range [0, 1]
    float resonance = 0.5f;  // range [0, 1]
    float mix = 1f;  // range [0, 1]
    };

    MbFilterAutowah() = default;
    ~MbFilterAutowah() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.sensitivity = std::clamp(params.sensitivity, 0f, 1f);
        params.speed = std::clamp(params.speed, 0f, 1f);
        params.depth = std::clamp(params.depth, 0f, 1f);
        params.resonance = std::clamp(params.resonance, 0f, 1f);
        params.mix = std::clamp(params.mix, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Auto-Wah
        return input;
    }
};

#endif // MB_FILTER_AUTOWAH_H

File: built-in plugins dsp/mb-filter-autowah.ts
import type { PluginDefinition } from "../server/services/pluginHostService";

const MbFilterAutowahPlugin: PluginDefinition = {
  id: "mb-filter-autowah",
  slug: "mb-filter-autowah",
  name: "MB Auto-Wah",
  category: "effect",
  type: "eq" as any,
  version: "1.0.0",
  description: "Envelope-following auto-wah effect",
  author: "Max Booster",
  grade: "A",
  parameters: [
    {
      id: "sensitivity",
      name: "Sensitivity",
      type: "float",
      defaultValue: 0.5,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
    {
      id: "speed",
      name: "Speed",
      type: "float",
      defaultValue: 0.5,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
    {
      id: "depth",
      name: "Depth",
      type: "float",
      defaultValue: 0.7,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
    {
      id: "resonance",
      name: "Resonance",
      type: "float",
      defaultValue: 0.5,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
    {
      id: "mix",
      name: "Mix",
      type: "float",
      defaultValue: 1,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
  ],
  defaultPreset: {
    sensitivity: 0.5,
    speed: 0.5,
    depth: 0.7,
    resonance: 0.5,
    mix: 1,
  },
};

export default MbFilterAutowahPlugin;
